avaframe/com1DFA/DFAfunctionsCython.pyx

```diff
@@ -1991,3 +2037,116 @@ cpdef double SamosATfric(double rho, double Rs0, double mu, double kappa, double
+  cdef double[:, :] nyArray = dem['Ny']
+  cdef double[:, :] nzArray = dem['Nz']
+  cdef double[:, :] areaRatser = dem['areaRaster']
+  cdef double[:] mass = particles['m']
+  cdef double[:] hArray = particles['h']
+  cdef double[:] xArray = particles['x']
+  cdef double[:] yArray = particles['y']
+  cdef double[:] zArray = particles['z']
+  cdef double[:] uxArray = particles['ux']
+  cdef double[:] uyArray = particles['uy']
+  cdef double[:] uzArray = particles['uz']
+  cdef int[:] indXDEM = particles['indXDEM']
+  cdef int[:] indYDEM = particles['indYDEM']
+  cdef double[:, :] VX = fields['Vx']
+  cdef double[:, :] VY = fields['Vy']
+  cdef double[:, :] VZ = fields['Vz']
+  cdef double[:] forceX = np.zeros(nPart, dtype=np.float64)
+  cdef double[:] forceY = np.zeros(nPart, dtype=np.float64)
+  cdef double[:] forceZ = np.zeros(nPart, dtype=np.float64)
+  cdef double[:] forceFrict = np.zeros(nPart, dtype=np.float64)
+  cdef double[:] dM = np.zeros(nPart, dtype=np.float64)
+
+  cdef int indCellX, indCellY
+  cdef double areaPart, uMag, m, h
+  cdef double vMeanx, vMeany, vMeanz, vMeanNorm, dvX, dvY, dvZ
+  cdef double x, y, z, xEnd, yEnd, zEnd, ux, uy, uz, uxDir, uyDir, uzDir
+  cdef double nx, ny, nz, nxEnd, nyEnd, nzEnd, nxAvg, nyAvg, nzAvg
+  # variables for interpolation
+  cdef int Lx0, Ly0, LxEnd0, LyEnd0, iCell, iCellEnd
+  cdef double w[4]
+  cdef double wEnd[4]
+  cdef int j
+  force = {}
+
+  for j in range(nPart):
+      m = mass[j]
+      x = xArray[j]
+      y = yArray[j]
+      z = zArray[j]
+      h = hArray[j]
+      ux = uxArray[j]
+      uy = uyArray[j]
+      uz = uzArray[j]
+      indCellX = indXDEM[j]
+      indCellY = indYDEM[j]
+
+      # deduce area
+      areaPart = m / (h * rho)
+      # get cell and weights
+      Lx0, Ly0, iCell, w[0], w[1], w[2], w[3] = getCellAndWeights(x, y, ncols, nrows, csz, interpOption)
+
+      # get normal at the particle location
+      nx, ny, nz = getVector(Lx0, Ly0, w[0], w[1], w[2], w[3], nxArray, nyArray, nzArray)
+      nx, ny, nz = normalize(nx, ny, nz)
+
+      # add artificial viscosity
+      vMeanx, vMeany, vMeanz = getVector(Lx0, Ly0, w[0], w[1], w[2], w[3], VX, VY, VZ)
+      # compute normal component of the velocity
+      vMeanNorm = scalProd(vMeanx, vMeany, vMeanz, nx, ny, nz)
+      # remove normal component (make sure vMean is in the tangent plane)
+      vMeanx = vMeanx - vMeanNorm * nx
+      vMeany = vMeany - vMeanNorm * ny
+      vMeanz = vMeanz - vMeanNorm * nz
+      # compute particle to field velocity difference
+      dvX = vMeanx - ux
+      dvY = vMeany - uy
+      dvZ = vMeanz - uz
+      dvMag = norm(dvX, dvY, dvZ)
+      Alat = 2.0 * np.sqrt((m * h) / rho)
+      fDrag = (subgridMixingFactor * 0.5 * rho * dvMag * Alat * dt) / m
+
+      # update velocity with artificial viscosity - implicit method
+      ux = ux + fDrag * vMeanx
+      uy = uy + fDrag * vMeany
+      uz = uz + fDrag * vMeanz
+      ux = ux / (1.0 + fDrag)
```

Comment:
Use the addArtificialviscosity function here

Reply:
done